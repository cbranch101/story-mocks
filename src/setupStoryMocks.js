import buildGetStoryProvider from './buildGetStoryProvider'
import setupTestWiringBase from './setupTestWiring'
import getSetupDecorator from './getSetupDecorator'
import getDecoratorWrapper from './getDecoratorWrapper'

const delay = ms => {
  return new Promise(resolve => setTimeout(resolve, ms))
}

const setupStoryMocks = ({storyWrappers = [], mapResults}) => {
  let currentFunctions = null

  const wrapApi = functions => {
    currentFunctions = functions
    return functions
  }

  const mock = (mockedFunctionsBase, options = {}) => {
    if (currentFunctions === null) {
      throw new Error(
        'Be sure to call wrapApi on your api before loading StoryMock stories or tests',
      )
    }
    const {
      onMockValueReturned = () => {},
      onMocksCreated = () => {},
      mapResults = results => results,
    } = options
    const mockedFunctions =
      typeof mockedFunctionsBase === 'function'
        ? mockedFunctionsBase(currentFunctions)
        : mockedFunctionsBase
    const delayAmount = mockedFunctions.DELAY_AMOUNT
    onMocksCreated()
    Object.keys(mockedFunctions)
      .filter(funcName => funcName !== 'DELAY_AMOUNT')
      .forEach(funcName => {
        currentFunctions[funcName] = async (...args) => {
          const getReturnValue = () => {
            const mockValue = mockedFunctions[funcName]
            if (typeof mockValue === 'function') {
              return mockValue(...args)
            }
            return mockValue
          }

          const returned = mapResults(getReturnValue())

          const isObject = typeof returned === 'object' && returned !== null

          const isError =
            isObject &&
            !!returned.TEST_ERROR &&
            Object.keys(returned).length === 1
          const resolveFunc = () =>
            isError
              ? Promise.reject(returned.TEST_ERROR)
              : Promise.resolve(returned)

          if (delayAmount) {
            await delay(delayAmount)
          }

          onMockValueReturned(funcName, ...args)
          return resolveFunc()
        }
      })
  }

  const getStoryProvider = buildGetStoryProvider(mock)
  const wrapper = getDecoratorWrapper({
    getStoryProvider,
    storyWrappers,
    mapResults,
  })
  const setupDecorator = getSetupDecorator({
    wrapper,
  })

  const setupTestWiring = (options = {}) => {
    const {mappedArgs, storyWrappers} = options
    return setupTestWiringBase({
      storyWrappers,
      mapResults,
      api: currentFunctions,
      getStoryProvider,
      mappedArgs,
    })
  }

  return {
    setupDecorator,
    wrapApi,
    setupTestWiring,
  }
}

export default setupStoryMocks
