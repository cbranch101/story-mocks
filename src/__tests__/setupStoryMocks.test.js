import React, {useEffect, useState} from 'react'
import {getRender} from 'react-wiring-library'
import setupStoryMocks from '../setupStoryMocks'

const {wrapApi, setupTestWiring} = setupStoryMocks({
  storyWrappers: [],
})

const baseApi = {
  getResponse: () => Promise.resolve('real-response'),
}

const api = wrapApi(baseApi)

const {wrapRender, getGlobalFunctions} = setupTestWiring([])

const DataFetcher = () => {
  const [response, setResponse] = useState([])
  useEffect(() => {
    const getResponse = async () => {
      try {
        const response = await api.getResponse('passed-into-response')
        setResponse(response)
      } catch (e) {
        setResponse(`Error: ${e}`)
      }
    }
    getResponse()
  }, [])

  const renderedResponse =
    typeof response === 'string' ? response : JSON.stringify(response)
  return <div data-testid="response">{renderedResponse}</div>
}

const wiring = {
  children: {
    response: {
      findValue: 'response',
      serialize: val => val.textContent,
    },
  },
}

const Story = () => <DataFetcher />

const setApiInStory = api => {
  Story.story = {
    parameters: {
      mocks: {
        api,
      },
    },
  }
}

const render = getRender(wiring, {
  render: wrapRender,
  customFunctions: {
    global: getGlobalFunctions,
  },
})

describe('setupStoryMocks helper', () => {
  describe('when the value of mocked api function is', () => {
    describe('a standard object', () => {
      test('that mocked object should be returned', async () => {
        setApiInStory({getResponse: 'mocked-response'})
        const {findResponse, waitForGetResponse} = render(Story)
        const args = await waitForGetResponse()
        expect(args).toMatchSnapshot('args passed into getResponse')
        const {response} = await findResponse()
        expect(response).toMatchSnapshot('initial render')
      })
    })
    describe('a function', () => {
      test('the result of calling that function should be returned', async () => {
        setApiInStory({getResponse: passedIn => `${passedIn}-extra-stuff`})
        const {findResponse} = render(Story)
        const {response} = await findResponse()
        expect(response).toMatchSnapshot('initial render')
      })
    })
    describe('an object with a single key of TEST_ERROR', () => {
      test('should cause the promise to reject with the provided error message', async () => {
        setApiInStory({getResponse: {TEST_ERROR: 'error-message'}})
        const {findResponse} = render(Story)
        const {response} = await findResponse()
        expect(response).toMatchSnapshot('initial render')
      })
    })
    describe('an object with a key of TEST_ERROR and other keys', () => {
      test('should return the response like normal', async () => {
        setApiInStory({
          getResponse: {TEST_ERROR: 'error-message', otherKey: true},
        })
        const {findResponse} = render(Story)
        const {response} = await findResponse()
        expect(response).toMatchSnapshot('initial render')
      })
    })
  })
  describe('when the entire passed in api is a function', () => {
    test('the return value of the function should behave like a normal api', async () => {
      setApiInStory(() => ({getResponse: 'result-of-a-func-call'}))
      const {findResponse} = render(Story)
      const {response} = await findResponse()
      expect(response).toMatchSnapshot('initial render')
    })
  })
  describe('when DELAY_AMOUNT is passed as a key to the api', () => {
    test('all api requests should be delayed by that amount', async () => {
      setApiInStory({getResponse: 'mocked-response', DELAY_AMOUNT: 100})
      const {findResponse, waitForGetResponse} = render(Story)
      await findResponse()
      const time = Date.now()
      await waitForGetResponse()
      const elapsed = Date.now() - time
      expect(elapsed).toBeGreaterThan(100)
    })
  })
})
