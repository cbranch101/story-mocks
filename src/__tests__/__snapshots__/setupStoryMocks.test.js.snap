// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupStoryMocks helper When a story with no mocks is tested the real api should be used: initial render 1`] = `real-response`;

exports[`setupStoryMocks helper When a story with no parameters is tested the real api should be used: initial render 1`] = `real-response`;

exports[`setupStoryMocks helper when a Story with no context is tested the real api should be used: initial render 1`] = `real-response`;

exports[`setupStoryMocks helper when a waitFor function is called multiple times should wait for the next call of the api: first call args 1`] = `
Array [
  "passed-into-response",
]
`;

exports[`setupStoryMocks helper when a waitFor function is called multiple times should wait for the next call of the api: second call args 1`] = `
Array [
  "second-call",
]
`;

exports[`setupStoryMocks helper when loading stories in the storybook environment mocks should behave the same as in tests: initial render 1`] = `mocked-response`;

exports[`setupStoryMocks helper when mapResults is not passed when the entire passed in api is a function the return value of the function should behave like a normal api: initial render 1`] = `result-of-a-func-call`;

exports[`setupStoryMocks helper when mapResults is not passed when the value of mocked api function is a function the result of calling that function should be returned: initial render 1`] = `passed-into-response-extra-stuff`;

exports[`setupStoryMocks helper when mapResults is not passed when the value of mocked api function is a standard object that mocked object should be returned: args passed into getResponse 1`] = `
Array [
  "passed-into-response",
]
`;

exports[`setupStoryMocks helper when mapResults is not passed when the value of mocked api function is a standard object that mocked object should be returned: initial render 1`] = `mocked-response`;

exports[`setupStoryMocks helper when mapResults is not passed when the value of mocked api function is an object with a key of TEST_ERROR and other keys should return the response like normal: initial render 1`] = `{"TEST_ERROR":"error-message","otherKey":true}`;

exports[`setupStoryMocks helper when mapResults is not passed when the value of mocked api function is an object with a single key of TEST_ERROR should cause the promise to reject with the provided error message: initial render 1`] = `Error: error-message`;

exports[`setupStoryMocks helper when mapResults is passed should call map results to update all returned mock results: initial render 1`] = `{"data":"mocked-response"}`;

exports[`setupStoryMocks helper when mappedArgs is passed should call the provided function on the correct api call to process its arguments before returning them from waitFor function: processed args 1`] = `"passed-into-response-cleaned"`;

exports[`setupStoryMocks helper when no args are passed to setupStoryMocks should use reasonable defaults: initial render 1`] = `mocked-response`;
